src/modules/projectConfig/projectConfigFileReader.ts

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache License, Version 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+import { PathLike } from 'node:fs';
+import { CliFileSystem } from '../../utils/fs-bridge';
+import { ComponentConfig } from '../component';
+import { PackageConfig } from '../package';
+import { ProjectConfig } from './projectConfig';
+import {
+    DEFAULT_CONFIG_FILE_PATH,
+    DEFAULT_CONFIG_LOCKFILE_PATH,
+    DesiredConfigFileComponents,
+    DesiredConfigFilePackages,
+    DesiredConfigFileVariables,
+    VARIABLE_SCOPE_SEPARATOR,
+} from './projectConfigConstants';
+import { ProjectConfigLock } from './projectConfigLock';
+
+interface IProjectConfigReader {
+    read(cliVersion: string, path: PathLike, ignoreLock: boolean): ProjectConfig;
+}
+
+export class MultiFormatConfigReader implements IProjectConfigReader {
```

Comment:
I suggest to avoid boilerplate code for instantiation of the correct reader/writer for ProjectConfig, we introduce a lean syntactic sugar class to do that for us: `ProjectConfigIO`.

If we ever swap Readers or Writers later, we will only have to do so in a central location.

```Typescript
class ProjectConfigIO {
   read(path: string): ProjectConfig {
      return (new MultiformatConfigReader()).read(path);
   }
   
   write(path: string, config: ProjectConfig) {
       (new ProjectConfigWriter()).write(config, path);
   }
}
```